packages/beacon-node/src/chain/shufflingCache.ts

```diff
@@ -152,39 +120,94 @@ export class ShufflingCache {
    */
-  async get(shufflingEpoch: Epoch, decisionRootHex: RootHex): Promise<EpochShuffling | null> {
-    const cacheItem = this.itemsByDecisionRootByEpoch.getOrDefault(shufflingEpoch).get(decisionRootHex);
+  async get(epoch: Epoch, decisionRoot: RootHex): Promise<EpochShuffling | null> {
+    const cacheItem = this.itemsByDecisionRootByEpoch.getOrDefault(epoch).get(decisionRoot);
     if (cacheItem === undefined) {
+      this.metrics?.shufflingCache.miss.inc();
       return null;
     }
 
     if (isShufflingCacheItem(cacheItem)) {
+      this.metrics?.shufflingCache.hit.inc();
       return cacheItem.shuffling;
     } else {
-      // promise
+      this.metrics?.shufflingCache.shufflingPromiseNotResolved.inc();
       return cacheItem.promise;
     }
   }
 
   /**
-   * Same to get() function but synchronous.
+   * Gets a cached shuffling via the epoch and decision root.  If the shuffling is not
+   * available it will build it synchronously and return the shuffling.
+   *
+   * NOTE: If a shuffling is already queued and not calculated it will build and resolve
+   * the promise but the already queued build will happen at some later time
    */
-  getSync(shufflingEpoch: Epoch, decisionRootHex: RootHex): EpochShuffling | null {
-    const cacheItem = this.itemsByDecisionRootByEpoch.getOrDefault(shufflingEpoch).get(decisionRootHex);
-    if (cacheItem === undefined) {
-      return null;
+  getSync<T extends ShufflingBuildProps | undefined>(
+    epoch: Epoch,
+    decisionRoot: RootHex,
+    buildProps?: T
+  ): T extends ShufflingBuildProps ? EpochShuffling : EpochShuffling | null {
+    const cacheItem = this.itemsByDecisionRootByEpoch.getOrDefault(epoch).get(decisionRoot);
+    if (!cacheItem) {
+      this.metrics?.shufflingCache.miss.inc();
+    } else if (isShufflingCacheItem(cacheItem)) {
+      this.metrics?.shufflingCache.hit.inc();
+      return cacheItem.shuffling;
+    } else if (buildProps) {
+      // TODO: (@matthewkeil) This should possible log a warning??
+      this.metrics?.shufflingCache.shufflingPromiseNotResolvedAndThrownAway.inc();
+    } else {
+      this.metrics?.shufflingCache.shufflingPromiseNotResolved.inc();
     }
 
-    if (isShufflingCacheItem(cacheItem)) {
-      return cacheItem.shuffling;
+    let shuffling: EpochShuffling | null = null;
+    if (buildProps) {
+      const timer = this.metrics?.shufflingCache.shufflingCalculationTime.startTimer();
+      shuffling = computeEpochShuffling(buildProps.state, buildProps.activeIndices, epoch);
+      timer?.();
+      this.set(shuffling, decisionRoot);
     }
+    return shuffling as T extends ShufflingBuildProps ? EpochShuffling : EpochShuffling | null;
+  }
 
-    // ignore promise
-    return null;
+  /**
+   * Queue asynchronous build for an EpochShuffling
+   */
+  build(epoch: number, decisionRoot: string, state: BeaconStateAllForks, activeIndices: ValidatorIndex[]): void {
```

Comment:
```suggestion
  build(epoch: number, decisionRoot: string, state: BeaconStateAllForks, activeIndices: Uint32Array): void {
```
this saves us 1 memory allocation